zerver/lib/actions.py

```diff
@@ -847,6 +866,40 @@ def active_humans_in_realm(realm: Realm) -> Sequence[UserProfile]:
     return UserProfile.objects.filter(realm=realm, is_active=True, is_bot=False)
 
 
+@transaction.atomic(savepoint=False)
+def update_users_in_full_members_system_group(
+    realm: Realm, affected_user_ids: Sequence[int] = []
+) -> None:
+    full_members_system_group = UserGroup.objects.get(
+        realm=realm, name="@role:fullmembers", is_system_group=True
+    )
+    members_system_group = UserGroup.objects.get(
+        realm=realm, name="@role:members", is_system_group=True
+    )
+
+    full_member_group_users = set(
+        full_members_system_group.direct_members.filter(id__in=affected_user_ids)
+    )
+    member_group_users = set(members_system_group.direct_members.filter(id__in=affected_user_ids))
+
+    old_full_members = [
+        user
+        for user in full_member_group_users
+        if user.is_provisional_member or user.role != UserProfile.ROLE_MEMBER
```

Comment:
Hmm, what was our decision on whether the right model is that "the full members system group directly contains all users who are at least members and not provisional" vs. "the full members system group contains all users whose role is member and are not provisional"?  

I guess they should be equivalent for all intents and purposes other than display (since the full members group contains the moderators group), and if we change our mind, the cron job will adjust within 5 minutes, so maybe it doesn't particularly matter. 

The other thought is that these queries are somewhat expensive and this cron job runs reasonably often, so I think it'd be worth refactoring the functions like `remove_members_from_user_group` to accept user IDs and have all the set work here just fetch a set of user IDs, not a set of UserProfile objects with 100 fields.

Reply:
We decided that full members group directly contains all users with role as member and are not provisional.

Also, regarding the refactoring, we also need role. Also to check `is_provisional_member` we would need to create a function which will use `role` and `date_joined` fields to find whether user is provisional or not, since we cannot access `is_provisional_member` if we instead use `values()` to get limited fields.

Reply:
OK, I think this sounds complex enough it might be better as a follow-up PR, since it's just a performance optimization.